src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp

```diff
@@ -0,0 +1,175 @@
+//
+// Note that we compute the product using derivatives 0...N in each dimension.
+// We then subtract off the term with 0 derivatives in all dimensions, because
+// that term is just the original data and should be ommitted.
+//
+// Note also that because the mean value of the data does not contribute to its
+// oscillation content, any matrix element associated with basis functions
+// m == 0 or n == 0 vanishes. We thus slightly reduce the necessary storage by
+// computing only the (N-1)^2 matrix where we start at m, n >= 1.
+template <size_t VolumeDim>
+Matrix compute_indicator_matrix(const Mesh<VolumeDim>& mesh) noexcept {
+  ASSERT(mesh.basis() == make_array<VolumeDim>(Spectral::Basis::Legendre),
+         "No implementation for mesh: " << mesh);
+  Matrix result(mesh.number_of_grid_points() - 1,
+                mesh.number_of_grid_points() - 1);
+  for (IndexIterator<VolumeDim> m(mesh.extents()); m; ++m) {
+    if (m.collapsed_index() == 0) {
+      // Skip the terms associated with the cell average of the data
+      continue;
+    }
+    for (IndexIterator<VolumeDim> n(mesh.extents()); n; ++n) {
+      if (n.collapsed_index() == 0) {
+        // Skip the terms associated with the cell average of the data
+        continue;
+      }
+      // Compute each indicator matrix term by tensor-product of 1D Legendre
+      // polynomial math
+      auto& result_mn =
+          result(m.collapsed_index() - 1, n.collapsed_index() - 1);
+      result_mn =
+          compute_sum_of_legendre_derivs(mesh.extents(0), m()[0], n()[0]);
+      for (size_t dim = 1; dim < VolumeDim; ++dim) {
+        result_mn *= compute_sum_of_legendre_derivs(mesh.extents(dim), m()[dim],
+                                                    n()[dim]);
+      }
+    }
+    // Subtract the tensor-product term that has no derivatives
+    double term_with_no_derivs =
+        Spectral::compute_basis_function_normalization_square<
+            Spectral::Basis::Legendre>(m()[0]);
+    for (size_t dim = 1; dim < VolumeDim; ++dim) {
+      term_with_no_derivs *=
+          Spectral::compute_basis_function_normalization_square<
+              Spectral::Basis::Legendre>(m()[dim]);
+    }
+    result(m.collapsed_index() - 1, m.collapsed_index() - 1) -=
+        term_with_no_derivs;
+  }
+  return result;
+}
+
+}  // namespace
+
+namespace SlopeLimiters {
+namespace Weno_detail {
+
+template <size_t VolumeDim>
+double oscillation_indicator(const DataVector& data,
+                             const Mesh<VolumeDim>& mesh) noexcept {
+  ASSERT(mesh.basis() == make_array<VolumeDim>(Spectral::Basis::Legendre),
+         "No implementation for mesh: " << mesh);
+
+  // Optimization: store the indicator matrix in a static so it is only computed
+  // once at the beginning of the simulation. This caching will have to be
+  // generalized to handle simulations with more than one mesh.
+  static const Matrix indicator_matrix = compute_indicator_matrix(mesh);
+
```

Comment:
Please then add an ASSERT that `indicator_matrix` is compatible with `mesh` so we don't forget this when there is more than one mesh